toolchain/check/import_ref.cpp

```diff
@@ -0,0 +1,266 @@
+    auto orig_fields = import_ir_.inst_blocks().Get(inst.fields_id);
+    llvm::SmallVector<SemIR::ConstantId> field_const_ids;
+    field_const_ids.reserve(orig_fields.size());
+    for (auto field_id : orig_fields) {
+      auto field = import_ir_.insts().GetAs<SemIR::StructTypeField>(field_id);
+      auto field_const_id = GetConstantId(field.field_type_id);
+      if (field_const_id.is_valid()) {
+        field_const_ids.push_back(field_const_id);
+      } else {
+        added_todos = true;
+      }
+    }
+    if (added_todos) {
+      return SemIR::ConstantId::Invalid;
+    }
+
+    // Prepare a vector of fields for GetStructType.
+    // TODO: Should we have field constants so that we can deduplicate fields
+    // without creating instructions here?
+    llvm::SmallVector<SemIR::InstId> fields;
+    fields.reserve(orig_fields.size());
+    for (auto [field_id, field_const_id] :
+         llvm::zip(orig_fields, field_const_ids)) {
+      auto field = import_ir_.insts().GetAs<SemIR::StructTypeField>(field_id);
+      auto name_str = import_ir_.names().GetAsStringIfIdentifier(field.name_id);
+      auto name_id = name_str ? SemIR::NameId::ForIdentifier(
+                                    context_.identifiers().Add(*name_str))
+                              : field.name_id;
+      auto field_type_id = context_.GetTypeIdForTypeConstant(field_const_id);
+      fields.push_back(context_.AddInstInNoBlock(
+          {Parse::NodeId::Invalid,
+           SemIR::StructTypeField{.name_id = name_id,
+                                  .field_type_id = field_type_id}}));
+    }
+
+    return context_.types().GetConstantId(
+        context_.GetStructType(context_.inst_blocks().Add(fields)));
+  }
+
+  auto TryResolveTypedInst(SemIR::TupleType inst) -> SemIR::ConstantId {
+    CARBON_CHECK(inst.type_id == SemIR::TypeId::TypeType);
+
+    // First verify that all used types are ready.
+    bool added_todos = false;
+    auto orig_el_type_ids = import_ir_.type_blocks().Get(inst.elements_id);
+    llvm::SmallVector<SemIR::ConstantId> el_const_ids;
+    el_const_ids.reserve(orig_el_type_ids.size());
+    for (auto el_type_id : orig_el_type_ids) {
+      auto el_const_id = GetConstantId(el_type_id);
+      if (el_const_id.is_valid()) {
+        el_const_ids.push_back(el_const_id);
+      } else {
+        added_todos = true;
+      }
+    }
+    if (added_todos) {
+      return SemIR::ConstantId::Invalid;
+    }
+
+    // Prepare a vector of the tuple types for GetTupleType.
+    llvm::SmallVector<SemIR::TypeId> el_type_ids;
+    el_type_ids.reserve(orig_el_type_ids.size());
+    for (auto el_const_id : el_const_ids) {
+      el_type_ids.push_back(context_.GetTypeIdForTypeConstant(el_const_id));
+    }
+
+    return context_.types().GetConstantId(context_.GetTupleType(el_type_ids));
+  }
+
+  Context& context_;
+  const SemIR::File& import_ir_;
+  SemIR::ConstantValueStore& import_ir_constant_values_;
+  llvm::SmallVector<SemIR::InstId> todo_;
```

Comment:
Particularly given the use of a to-do API elsewhere, id prefer a different name. My canonical name for this pattern is "worklist".

Reply:
(as discussed, let's go with `work_stack_`)

Reply:
Done